From b2530e2214ad84a5606042a9003993429848ba6d Mon Sep 17 00:00:00 2001
From: =?UTF-8?q?Ond=C5=99ej=20Caletka?= <ondrej@caletka.cz>
Date: Tue, 11 Sep 2012 14:48:48 +0200
Subject: [PATCH 4/5] Do not trust AD flag when using ldns

Instead of relying on (possibly spoofed) AD flag from upstream DNS
server, try to validate DNSSEC autonomously.
For this to work, the trust anchor have to be configured in /etc/resolv.conf:
anchor /path/to/root.key
---
 openbsd-compat/getrrsetbyname-ldns.c | 43 +++++++++++++++++-------------------
 1 file changed, 20 insertions(+), 23 deletions(-)

diff --git a/openbsd-compat/getrrsetbyname-ldns.c b/openbsd-compat/getrrsetbyname-ldns.c
index 1966634..b89d2e9 100644
--- a/openbsd-compat/getrrsetbyname-ldns.c
+++ b/openbsd-compat/getrrsetbyname-ldns.c
@@ -148,32 +148,29 @@ getrrsetbyname(const char *hostname, unsigned int rdclass,
 	debug2("ldns: got %u answers from DNS", rrset->rri_nrdatas);
 
 	/* Check for authenticated data */
-	if (ldns_pkt_ad(pkt)) {
-		rrset->rri_flags |= RRSET_VALIDATED;
-	} else { /* AD is not set, try autonomous validation */
-		ldns_rr_list * trusted_keys = ldns_rr_list_new();
-
-		debug2("ldns: trying to validate RRset");
-		/* Get eventual sigs */
-		rrsigs = ldns_pkt_rr_list_by_type(pkt, LDNS_RR_TYPE_RRSIG,
-		    LDNS_SECTION_ANSWER);
-
-		rrset->rri_nsigs = ldns_rr_list_rr_count(rrsigs);
-		debug2("ldns: got %u signature(s) (RRTYPE %u) from DNS",
-		       rrset->rri_nsigs, LDNS_RR_TYPE_RRSIG);
-
-		if ((err = ldns_verify_trusted(ldns_res, rrdata, rrsigs,
-		     trusted_keys)) == LDNS_STATUS_OK) {
-			rrset->rri_flags |= RRSET_VALIDATED;
-			debug2("ldns: RRset is signed with a valid key");
-		} else {
-			debug2("ldns: RRset validation failed: %s",
-			    ldns_get_errorstr_by_id(err));
-		}
+	/* try autonomous validation, regardless of AD flag */
+	ldns_rr_list * trusted_keys = ldns_rr_list_new();
+
+	debug2("ldns: trying to validate RRset");
+	/* Get eventual sigs */
+	rrsigs = ldns_pkt_rr_list_by_type(pkt, LDNS_RR_TYPE_RRSIG,
+	    LDNS_SECTION_ANSWER);
 
-		ldns_rr_list_deep_free(trusted_keys);
+	rrset->rri_nsigs = ldns_rr_list_rr_count(rrsigs);
+	debug2("ldns: got %u signature(s) (RRTYPE %u) from DNS",
+	       rrset->rri_nsigs, LDNS_RR_TYPE_RRSIG);
+
+	if ((err = ldns_verify_trusted(ldns_res, rrdata, rrsigs,
+	     trusted_keys)) == LDNS_STATUS_OK) {
+		rrset->rri_flags |= RRSET_VALIDATED;
+		debug2("ldns: RRset is signed with a valid key");
+	} else {
+		debug2("ldns: RRset validation failed: %s",
+		    ldns_get_errorstr_by_id(err));
 	}
 
+	ldns_rr_list_deep_free(trusted_keys);
+
 	/* allocate memory for answers */
 	rrset->rri_rdatas = calloc(rrset->rri_nrdatas,
 	   sizeof(struct rdatainfo));
-- 
1.7.12.4

